syntax = "proto3";

import "common.proto";
import "enums.proto";

service WorkerToScheduler {
    // Registers the worker with the scheduler
    rpc RegisterWorker (RegisterWorkerRequest) returns (RegisterWorkerResponse);

    // Heartbeat indicating worker status
    rpc SendHeartbeat (Heartbeat) returns (Empty);

    // Indicates to the scheduler that a job has completed
    rpc Done (DoneRequest) returns (Empty);
}

message RegisterWorkerRequest {
    message Device {
        uint32 device_id = 1;
        DeviceType device_type = 2;
        float available_memory = 3;
        }
    string worker_type = 1;
    string ip_addr = 2;
    uint32 port = 3;
    repeated Device devices = 4;
}

message RegisterWorkerResponse {
    bool success = 1;
    uint64 worker_id = 2;
    uint64 round_duration = 3;
    string error_message = 4;
}

message Heartbeat {
    uint64 worker_id = 1;
    repeated JobState job_state = 2;
}

message DoneRequest {
    uint64 worker_id = 1;
    repeated uint64 job_id = 2;
    repeated uint64 num_steps = 3;
    repeated double execution_time = 4;
}
